File: src/extractors.ts
import fs from 'fs';
import Papa from 'papaparse';
import { Evidence } from './models';

const KNOWN_SKILLS = [
  'python',
  'java',
  'javascript',
  'sql',
  'aws',
  'docker',
  'kubernetes',
  'machine learning',
  'nlp',
  'react',
  'node',
  'postgres',
];

export function extractFromRecruiterCsv(filePath: string | null): Evidence[] {
  if (!filePath || !fs.existsSync(filePath)) return [];

  const result: Evidence[] = [];
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    const parsed = Papa.parse<Record<string, string>>(content, {
      header: true,
      skipEmptyLines: true,
    });

    for (const row of parsed.data) {
      const name = row.name || null;
      const email = row.email || null;
      const phone = row.phone || null;
      const company = row.current_company || null;
      const title = row.title || null;
      const location = row.location || null;
      const skills = row.skills || null;

      if (name) result.push({ field: 'full_name', value: name, confidence: 0.93, source: 'recruiter_csv', method: 'column:name' });
      if (email) result.push({ field: 'emails', value: email, confidence: 0.97, source: 'recruiter_csv', method: 'column:email' });
      if (phone) result.push({ field: 'phones', value: phone, confidence: 0.9, source: 'recruiter_csv', method: 'column:phone' });
      if (title) result.push({ field: 'headline', value: title, confidence: 0.7, source: 'recruiter_csv', method: 'column:title' });

      if (company || title) {
        result.push({
          field: 'experience',
          value: { company, title, start: null, end: null, summary: 'Current role from recruiter export' },
          confidence: 0.75,
          source: 'recruiter_csv',
          method: 'row_to_experience',
        });
      }

      if (location) {
        const locParts = location.split(',').map((p: string) => p.trim());
        if (locParts[0]) result.push({ field: 'location.city', value: locParts[0], confidence: 0.8, source: 'recruiter_csv', method: 'parse_location' });
        if (locParts[1]) result.push({ field: 'location.region', value: locParts[1], confidence: 0.8, source: 'recruiter_csv', method: 'parse_location' });
        if (locParts[2]) result.push({ field: 'location.country', value: locParts[2], confidence: 0.8, source: 'recruiter_csv', method: 'parse_location' });
      }

      if (skills) {
        const skillList = skills.split(',').map((s: string) => s.trim());
        for (const skill of skillList) {
          if (skill) result.push({ field: 'skills', value: skill, confidence: 0.8, source: 'recruiter_csv', method: 'column:skills' });
        }
      }
    }
  } catch (e) {
    // ignore parsing errors
  }

  return result;
}

export function extractFromAtsJson(filePath: string | null): Evidence[] {
  if (!filePath || !fs.existsSync(filePath)) return [];

  const result: Evidence[] = [];
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    const payload = JSON.parse(content);

    if (payload.fullName) result.push({ field: 'full_name', value: payload.fullName, confidence: 0.88, source: 'ats_json', method: 'field:fullName' });
    if (payload.primaryEmail) result.push({ field: 'emails', value: payload.primaryEmail, confidence: 0.92, source: 'ats_json', method: 'field:primaryEmail' });
    if (payload.phoneNumber) result.push({ field: 'phones', value: payload.phoneNumber, confidence: 0.87, source: 'ats_json', method: 'field:phoneNumber' });
    if (payload.yearsExp !== undefined) result.push({ field: 'years_experience', value: payload.yearsExp, confidence: 0.78, source: 'ats_json', method: 'field:yearsExp' });

    if (payload.links && typeof payload.links === 'object') {
      if (payload.links.github) result.push({ field: 'links.github', value: payload.links.github, confidence: 0.83, source: 'ats_json', method: 'field:links.github' });
      if (payload.links.linkedin) result.push({ field: 'links.linkedin', value: payload.links.linkedin, confidence: 0.83, source: 'ats_json', method: 'field:links.linkedin' });
    }

    if (Array.isArray(payload.skills)) {
      for (const skill of payload.skills) {
        result.push({ field: 'skills', value: skill, confidence: 0.85, source: 'ats_json', method: 'field:skills' });
      }
    }

    if (Array.isArray(payload.experience)) {
      for (const exp of payload.experience) {
        if (typeof exp === 'object') {
          result.push({ field: 'experience', value: exp, confidence: 0.8, source: 'ats_json', method: 'field:experience' });
        }
      }
    }

    if (Array.isArray(payload.education)) {
      for (const edu of payload.education) {
        if (typeof edu === 'object') {
          result.push({ field: 'education', value: edu, confidence: 0.82, source: 'ats_json', method: 'field:education' });
        }
      }
    }
  } catch (e) {
    // ignore parsing errors
  }

  return result;
}

export function extractFromNotesTxt(filePath: string | null): Evidence[] {
  if (!filePath || !fs.existsSync(filePath)) return [];

  const result: Evidence[] = [];
  try {
    const text = fs.readFileSync(filePath, 'utf-8');

    const nameMatch = text.match(/name\s*:\s*(.+)/i);
    if (nameMatch) result.push({ field: 'full_name', value: nameMatch[1].trim(), confidence: 0.72, source: 'notes_txt', method: 'regex:name' });

    const emails = text.match(/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g) || [];
    for (const email of emails) {
      result.push({ field: 'emails', value: email, confidence: 0.7, source: 'notes_txt', method: 'regex:email' });
    }

    const phones = text.match(/(?:\+?\d[\d\s().-]{8,}\d)/g) || [];
    for (const phone of phones) {
      result.push({ field: 'phones', value: phone, confidence: 0.65, source: 'notes_txt', method: 'regex:phone' });
    }

    const yearsMatch = text.match(/(\d+)\+?\s+years/i);
    if (yearsMatch) {
      result.push({ field: 'years_experience', value: parseInt(yearsMatch[1], 10), confidence: 0.7, source: 'notes_txt', method: 'regex:years_experience' });
    }

    const linkedin = text.match(/https?:\/\/(?:www\.)?linkedin\.com\/\S+/i);
    if (linkedin) result.push({ field: 'links.linkedin', value: linkedin[0], confidence: 0.74, source: 'notes_txt', method: 'regex:linkedin' });

    const github = text.match(/https?:\/\/(?:www\.)?github\.com\/\S+/i);
    if (github) result.push({ field: 'links.github', value: github[0], confidence: 0.74, source: 'notes_txt', method: 'regex:github' });

    for (const skill of KNOWN_SKILLS) {
      const skillRegex = new RegExp(`\\b${skill}\\b`, 'i');
      if (skillRegex.test(text)) {
        result.push({ field: 'skills', value: skill, confidence: 0.62, source: 'notes_txt', method: 'keyword_match' });
      }
    }
  } catch (e) {
    // ignore parsing errors
  }

  return result;
}

File: src/extractors.py
from __future__ import annotations

import csv
import json
import re
from pathlib import Path

from .models import Evidence

KNOWN_SKILLS = [
    "python",
    "java",
    "javascript",
    "sql",
    "aws",
    "docker",
    "kubernetes",
    "machine learning",
    "nlp",
    "react",
    "node",
    "postgres",
]


def extract_from_recruiter_csv(path: str | None) -> list[Evidence]:
    if not path:
        return []
    p = Path(path)
    if not p.exists():
        return []

    out: list[Evidence] = []
    with p.open("r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            name = row.get("name")
            email = row.get("email")
            phone = row.get("phone")
            company = row.get("current_company")
            title = row.get("title")
            location = row.get("location")
            skills = row.get("skills")

            if name:
                out.append(Evidence("full_name", name, 0.93, "recruiter_csv", "column:name"))
            if email:
                out.append(Evidence("emails", email, 0.97, "recruiter_csv", "column:email"))
            if phone:
                out.append(Evidence("phones", phone, 0.9, "recruiter_csv", "column:phone"))
            if title:
                out.append(Evidence("headline", title, 0.7, "recruiter_csv", "column:title"))
            if company or title:
                out.append(
                    Evidence(
                        "experience",
                        {
                            "company": company,
                            "title": title,
                            "start": None,
                            "end": None,
                            "summary": "Current role from recruiter export",
                        },
                        0.75,
                        "recruiter_csv",
                        "row_to_experience",
                    )
                )
            if location:
                parts = [x.strip() for x in location.split(",")]
                city = parts[0] if len(parts) > 0 else None
                region = parts[1] if len(parts) > 1 else None
                country = parts[2] if len(parts) > 2 else None
                if city:
                    out.append(Evidence("location.city", city, 0.8, "recruiter_csv", "parse_location"))
                if region:
                    out.append(Evidence("location.region", region, 0.8, "recruiter_csv", "parse_location"))
                if country:
                    out.append(Evidence("location.country", country, 0.8, "recruiter_csv", "parse_location"))
            if skills:
                for s in [k.strip() for k in skills.split(",") if k.strip()]:
                    out.append(Evidence("skills", s, 0.8, "recruiter_csv", "column:skills"))
    return out


def extract_from_ats_json(path: str | None) -> list[Evidence]:
    if not path:
        return []
    p = Path(path)
    if not p.exists():
        return []

    try:
        payload = json.loads(p.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return []

    out: list[Evidence] = []
    if name := payload.get("fullName"):
        out.append(Evidence("full_name", name, 0.88, "ats_json", "field:fullName"))
    if email := payload.get("primaryEmail"):
        out.append(Evidence("emails", email, 0.92, "ats_json", "field:primaryEmail"))
    if phone := payload.get("phoneNumber"):
        out.append(Evidence("phones", phone, 0.87, "ats_json", "field:phoneNumber"))
    if yrs := payload.get("yearsExp"):
        out.append(Evidence("years_experience", yrs, 0.78, "ats_json", "field:yearsExp"))
    if links := payload.get("links", {}):
        if isinstance(links, dict):
            if gh := links.get("github"):
                out.append(Evidence("links.github", gh, 0.83, "ats_json", "field:links.github"))
            if li := links.get("linkedin"):
                out.append(Evidence("links.linkedin", li, 0.83, "ats_json", "field:links.linkedin"))

    for sk in payload.get("skills", []):
        out.append(Evidence("skills", sk, 0.85, "ats_json", "field:skills"))

    for item in payload.get("experience", []):
        if isinstance(item, dict):
            out.append(Evidence("experience", item, 0.8, "ats_json", "field:experience"))

    for item in payload.get("education", []):
        if isinstance(item, dict):
            out.append(Evidence("education", item, 0.82, "ats_json", "field:education"))
    return out


def extract_from_notes_txt(path: str | None) -> list[Evidence]:
    if not path:
        return []
    p = Path(path)
    if not p.exists():
        return []

    text = p.read_text(encoding="utf-8", errors="ignore")
    out: list[Evidence] = []

    name_match = re.search(r"name\s*:\s*(.+)", text, flags=re.IGNORECASE)
    if name_match:
        out.append(Evidence("full_name", name_match.group(1).strip(), 0.72, "notes_txt", "regex:name"))

    for email in re.findall(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}", text):
        out.append(Evidence("emails", email, 0.7, "notes_txt", "regex:email"))

    for phone in re.findall(r"(?:\+?\d[\d\s().-]{8,}\d)", text):
        out.append(Evidence("phones", phone, 0.65, "notes_txt", "regex:phone"))

    years_match = re.search(r"(\d+)\+?\s+years", text, flags=re.IGNORECASE)
    if years_match:
        out.append(
            Evidence(
                "years_experience",
                int(years_match.group(1)),
                0.7,
                "notes_txt",
                "regex:years_experience",
            )
        )

    linkedin = re.search(r"https?://(?:www\.)?linkedin\.com/\S+", text, flags=re.IGNORECASE)
    if linkedin:
        out.append(Evidence("links.linkedin", linkedin.group(0), 0.74, "notes_txt", "regex:linkedin"))

    github = re.search(r"https?://(?:www\.)?github\.com/\S+", text, flags=re.IGNORECASE)
    if github:
        out.append(Evidence("links.github", github.group(0), 0.74, "notes_txt", "regex:github"))

    for skill in KNOWN_SKILLS:
        if re.search(rf"\b{re.escape(skill)}\b", text, flags=re.IGNORECASE):
            out.append(Evidence("skills", skill, 0.62, "notes_txt", "keyword_match"))

    return out
